### PR TITLE
[rllib] Fix incorrect log value of environment steps sampled/trained

### PR DESCRIPTION
**Pull request by Daraan**
## Why are these changes needed?

Currently `learners/__all_modules__/num_env_steps_trained` has a strangely large value, multiple times higher than `learners/__all_modules__/num_module_steps_trained`.
This is due to a logging error as `minibatch.env_steps()` reports the size of the whole batch.

This is confusing but, seems to be intended (once):
`minibatch = MultiAgentBatch(minibatch, env_steps=len(whole_batch))`

https://github.com/ray-project/ray/blob/3a1750b07f715390f15fa8f57b8bc879829ae2ec/rllib/utils/minibatch_utils.py#L174-L178

Due to this fallacy the *whole* batch_size is logged for each *mini* batch iteration creating a too large value:

---

For example:

I have a batch size and sample size of 2048, train for 20 epochs with a minibatch size of 128. The resulting learners log is:
```
__all_modules__: {
  num_module_steps_trained: 40960,
  num_env_steps_trained: 655360
}
```

`num_module_steps_trained` is correct
num_module_steps_trained = 2048 samples * 20 epochs = 128 minibatch_size * (2048/128 minibatch cycles) * 20 epochs.

However, `num_env_steps_trained` makes no sense - it is 16 times higher. It is calculated:
2048 samples * 20 epochs * (16 = 2048 / 128 minibatch cycles) = 2048 * 320 iterations total

## Related issue number

https://discuss.ray.io/t/is-the-num-env-steps-trained-logged-incorrectly-if-not-how-to-interpret-it-compared-to-num-module-steps-trained/22616

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests

---

> [!NOTE]
> Log environment steps using per-minibatch module size instead of whole-batch env_steps, aligning env/module step counters and add tests to verify aggregation.
> 
> - **Metrics logging (training/evaluation)**:
>   - Update `rllib/core/learner/learner.py` and `rllib/core/learner/differentiable_learner.py` to log `ALL_MODULES` `NUM_ENV_STEPS_TRAINED` and `NUM_ENV_STEPS_TRAINED_LIFETIME` using `module_batch_size` instead of `batch.env_steps()`.
>   - Update `rllib/offline/offline_evaluation_runner.py` to log `NUM_ENV_STEPS_SAMPLED` and `NUM_ENV_STEPS_SAMPLED_LIFETIME` using `module_batch_size`.
> - **Tests**:
>   - Extend `rllib/core/learner/tests/test_learner.py` with assertions tying `NUM_ENV_STEPS_*` to corresponding `NUM_MODULE_STEPS_*`, plus a new multi-agent aggregation test ensuring per-module and `ALL_MODULES` sums match.
>   - Extend `rllib/offline/tests/test_offline_evaluation_runner_group.py` with checks that sampled env-step metrics equal module-step metrics.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a6b69bc3fdae8f35705c4e10433efc783066dba0. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by pseudo-rnd-thoughts** [comment]
Thanks for the PR @Daraan 

Could you give an example to test with? As I understand for multi-agent settings the envs step, agent steps, and policy updates can differ significantly. 

Also could you simplify your testing code to
```python
self.assertEqual(
    results[DEFAULT_MODULE_ID][NUM_MODULE_STEPS_TRAINED].peek(),
    results[ALL_MODULES][NUM_ENV_STEPS_TRAINED].peek()
)
self.assertEqual(
    results[DEFAULT_MODULE_ID][NUM_MODULE_STEPS_TRAINED_LIFETIME].peek(),
    results[ALL_MODULES][NUM_ENV_STEPS_TRAINED_LIFETIME].peek(),
)
```

**Comment by Daraan** [comment]
I extended it with a test with multiple modules. I guess someone else can do a better follow up.

**Comment by Daraan** [comment]
@pseudo-rnd-thoughts any further thoughts. Give it a go, or do you potentially want to co-contribute with a fitting test covering more agent/envs? I guess you are more experienced how this could be set up.

**Comment by pseudo-rnd-thoughts** [comment]
@kamil-kaczmarek Is already investigating the metrics as we are aware of issues. I'll let him resolve it

**Comment by Daraan** [comment]
Took over some patches from main. Possibly wrong again as `env_steps` is back. Lets see what the test does need to familiarize myself again with the logic.

**Comment by pseudo-rnd-thoughts** [comment]
@Daraan Is this still needed if https://github.com/ray-project/ray/pull/56838 is merged?

**Comment by Daraan** [comment]
@pseudo-rnd-thoughts Yes, it is not related to the metrics system but to a logical difference in how `batch.env_steps()` is implemented and how we log it - see top post.
With https://github.com/ray-project/ray/pull/57215 I thought parts of my PR are superseded by it. But that was not true.

To summarize it the fundamental problem is that `batch.env_steps()` always returns the same value, using minibatches via `MiniBatchCyclicIterator` does *not* change the result. However, when we use `num_epochs` and `minibatch_size` in learners we iterate multiple times over the logging of `batch.env_steps()` - which results in a by *batch_size / minibatch_size* scaled up value which is wrong.

In my inital version of this PR I substituted the env_steps with the respective module_steps. While that seems to be okay in the test-suite and standard rllib it still gives me a bad vibe now as we kinda ignore the `env_steps=` parameter for the batch creation.

As noted it seems intentional that the `MiniBatchCyclicIterator` does this - but I do not see why.

https://github.com/ray-project/ray/blob/3a1750b07f715390f15fa8f57b8bc879829ae2ec/rllib/utils/minibatch_utils.py#L174-L178

---

But we do need to change this behavior somehow, in `_log_steps_trained_metrics` we do now know if the count of module steps is equivalent to the env steps - could be a single env reused mutliple times are multiple environments.

On the other hand I think scaling down the result of `minibatch_batch.env_steps()` itself is also not correct.

What we need is:

```python
 self.metrics.log_value(
            (ALL_MODULES, NUM_ENV_STEPS_TRAINED),
            batch.env_steps() / count_minibatches,
            reduce="sum",
            with_throughput=True,
        )
```

EDIT:

However we do not know count_minibatches beforehand. I think we need some kind of post update of the metric after we know it; or only update once for each epoch.

A simple possibility could be to limit the updates like this:

```python
            self._log_steps_trained_metrics(
                tensor_minibatch,
                update_env_steps=iteration < num_epochs,
            )
```
However, as the samples envs steps are normally *train_batch_size_per_learner + small overhead*, i.e. a count of 2050 is common when we just want 2048, there is one more loop than we might expect. NUM_ENV_STEPS_TRAINED will diverge from NUM_MODULE_STEPS_TRAINED for this single iteration.

@pseudo-rnd-thoughts any thoughts on this?

**Comment by PhilippWillms** [comment]
*push* - is this topic still occurring in 2.53 release ?

**Comment by ArturNiederfahrenhorst** [comment]
@Daraan Thanks for taking a look at this.
There was an assumption in this PR that does not hold since the new Stats and MetricsLogger:
When we get back lifetime stats from learners or envrunners, these don't represent all values collected over the lifetime from the respective component because they will be aggregated only at the root metrics logger.

I changed the PR to respect that but values still didn't really add up because of https://github.com/ray-project/ray/pull/56599/files#diff-6d4917313c4715edfb04b9797ea992e21ff974afabc5cbcac3c3c47bddfb1771L379
so I'm converting this PR a bit to one that tries to deal with that exact issue. Let's see how tests go.

**Comment by Daraan** [comment]
@ArturNiederfahrenhorst Lovely, thank you for taking a look at it

File: rllib/core/learner/tests/test_learner.py [hunk 1]
```diff
@@ -29,21 +38,48 @@ def tearDown(cls) -> None:
...
+            )
+            self.assertEqual(batch.count, results[ALL_MODULES][NUM_ENV_STEPS_TRAINED])
+            self.assertEqual(
+                batch.count, results[ALL_MODULES][NUM_ENV_STEPS_TRAINED_LIFETIME]
+            )
+
+        loss = results[DEFAULT_MODULE_ID][Learner.TOTAL_LOSS_KEY]
         min_loss = min(loss, min_loss)
-        print(f"[iter = {iter_i}] Loss: {loss:.3f}, Min Loss: {min_loss:.3f}")
+        print(f"[iter = {seq_num}] Loss: {loss:.3f}, Min Loss: {min_loss:.3f}")
```
Comment (pseudo-rnd-thoughts) [inline rllib/core/learner/tests/test_learner.py]: Should the test have this debug print statement?

File: rllib/core/learner/tests/test_learner.py [hunk 2]
```diff
@@ -29,21 +38,48 @@ def tearDown(cls) -> None:
...
+        """
 
         config = BaseTestingAlgorithmConfig()
 
         learner = config.build_learner(env=self.ENV)
         reader = get_cartpole_dataset_reader(batch_size=512)
 
         min_loss = float("inf")
-        for iter_i in range(1000):
+        for seq_num in range(1, 1000):
```
Comment (pseudo-rnd-thoughts) [inline rllib/core/learner/tests/test_learner.py:51]: Why is the sequence number not 0-indexed? Do we think of original learner state as 0, then the first updated state is 1 making it 1-indexed?

File: rllib/core/learner/tests/test_learner.py [hunk 3]
```diff
@@ -29,21 +38,48 @@ def tearDown(cls) -> None:
...
             batch = reader.next().as_multi_agent()
             batch = learner._convert_batch_type(batch)
             results = learner.update(batch=batch)
 
-        loss = results[DEFAULT_MODULE_ID][Learner.TOTAL_LOSS_KEY].peek()
+            self.assertEqual(
```
Comment (pseudo-rnd-thoughts) [inline rllib/core/learner/tests/test_learner.py:59]: Could there be better grouping of the metrics

| metric | reference | 
|---|---|
| Number of module steps trained | batch.count |
| Number of module steps trained lifetime | batch.count |
| Weight sequence number | Sequence Number |
| Module Train batch size mean | batch.count |
| All Module steps trained | batch.count | 
| All Module steps trained lifetime | batch.count |   
| All module num steps trained | batch.count | 
| All module num steps trained lifetime | batch.count | 

For the total loss, could we confirm the value is a scalar? 
For the lifetime metrics, I'm surprised for each learner update they are equivalent to the intermediate learner update, @ArturNiederfahrenhorst is this due to the lifetime not being compiled yet? If so, could we be testing the metric result here and not in the root metric?

File: rllib/core/learner/tests/test_learner.py [hunk 4]
```diff
@@ -239,6 +275,63 @@ def _check_learner_states(self, framework, learner1, learner2):
...
+                    results[module_id][NUM_MODULE_STEPS_TRAINED_LIFETIME],
+                )
+                self.assertEqual(seq_num, results[module_id][WEIGHTS_SEQ_NO])
+                self.assertEqual(
+                    sa_batch_count, results[module_id][MODULE_TRAIN_BATCH_SIZE_MEAN]
+                )
+                # We don't know what the value should be, just check for existence.
+                self.assertTrue(learner.TOTAL_LOSS_KEY in results[module_id])
+
+            self.assertEqual(
```
Comment (pseudo-rnd-thoughts) [inline rllib/core/learner/tests/test_learner.py:318]: Same question about lifetime metrics as above

File: rllib/core/learner/tests/test_learner.py [hunk 5]
```diff
@@ -29,21 +38,48 @@ def tearDown(cls) -> None:
...
         config = BaseTestingAlgorithmConfig()
 
         learner = config.build_learner(env=self.ENV)
         reader = get_cartpole_dataset_reader(batch_size=512)
 
         min_loss = float("inf")
-        for iter_i in range(1000):
+        for seq_num in range(1, 1000):
             batch = reader.next().as_multi_agent()
             batch = learner._convert_batch_type(batch)
```
Comment (pseudo-rnd-thoughts) [inline rllib/core/learner/tests/test_learner.py:53]: Add check that the batch size is equal to 512 as currently the tests can pass while all not matching the configurations

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]

File: rllib/core/learner/tests/test_learner.py (hunk 2 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/learner/tests/test_learner.py:51]: That's because we initialize seq_num with zero. Any update will update the sequnce number in the Learner with `self._weights_seq_no += 1`.

File: rllib/core/learner/tests/test_learner.py (hunk 1 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/learner/tests/test_learner.py]: I'm good with removing it! Removed.

File: rllib/core/learner/tests/test_learner.py (hunk 3 above)
Comment (ArturNiederfahrenhorst) [reply rllib/core/learner/tests/test_learner.py:59]: Yes, indeed. No parallel component has a global view of lifetime metrics. They therefore just send their contribution to any metrics (lifetime or not) downstream.

Comment (ArturNiederfahrenhorst) [reply rllib/core/learner/tests/test_learner.py:59]: > For the total loss, could we confirm the value is a scalar?

Since we are inspectinv the results dictionary here, the returned value is of type MeanStats.

Comment (ArturNiederfahrenhorst) [reply rllib/core/learner/tests/test_learner.py:59]: > If so, could we be testing the metric result here and not in the root metric?

That's indeed what's happening. Testing aggregation of such metrics happens elsewhere - this testing is component level. I think it is enough to returned metrics for each component and prove that aggregation (also over lifetime) works well in the Stats/MetricsLogger tests.

**Review by pseudo-rnd-thoughts — APPROVED** [review]
LGTM
